src/device/cuda_device.zig

```diff
@@ -0,0 +1,341 @@
+            cuda.memFree(@as(*anyopaque, @ptrFromInt(self.head)), stream);
+        }
+        self.* = undefined;
+    }
+    // Each device has it's own scratch memory because streams work
+    // like queues. It's safe if the same queue tries to access its
+    // own memory, but dangerous if streams can use other scratch.
+    pub fn get(self: *ScratchMemory, comptime T: type, n: usize, stream: *anyopaque) []T {
+        const total: usize = @sizeOf(T) * n;
+        // check if we have enough scratch to provide a payload
+        if (self.tail < (self.head + total)) {
+            if (self.head != 0) {
+                cuda.memFree(@as(*anyopaque, @ptrFromInt(self.head)), stream);
+            }
+            // after a first pass through the network, we should know if we have enough memory.
+            const ptr = cuda.memAlloc(total) orelse @panic("Cannot allocate scratch memory.");
+            self.head = @intFromPtr(ptr);
+            self.tail = self.head + total;
+        }
+        const ptr: [*]T = @ptrFromInt(self.scratch.head);
+        return ptr[0..n];
+    }
+};
+
+
+pub const CudaDevice = struct {
+
+    const Error = std.mem.Allocator.Error;
+
+    // keep these out of the user api
+    context: struct {
+        device_number: u32,
+        stream: *anyopaque,
+        cublas: *anyopaque,
+        cudnn: *anyopaque,
+    },
+    nn: NN,
+    blas: Blas,
+    cache: DataCache,
+    scratch: ScratchMemory,
+    allocator: std.mem.Allocator,
+
+    pub fn init(device_number: u32, backing_allocator: std.mem.Allocator) CudaDevice {
+        cuda.initDevice(device_number);
+        const stream = cuda.initStream() orelse unreachable;
+        const cublas = cuda.initCublasHandle(stream) orelse unreachable;
+        const cudnn = cuda.initCudnnHandle(stream) orelse unreachable;
+        return .{
+            .context = .{
+                .device_number = device_number,
+                .stream = stream,
+                .cublas = cublas,  
+                .cudnn = cudnn,
+            },
+            .nn = .{},
+            .blas = .{},
+            // at some point, maybe move this to unmanged
+            .cache = .{ .allocator = backing_allocator },
+            .scratch = .{},
+            .allocator = backing_allocator,
+        };
+    }
+
+    pub fn deinit(self: *CudaDevice) void {
+        // free everything we own first...
+        var iter = self.cache.poppingIterator();
+        while (iter.next()) |ptr| cuda.memFree(ptr, self.context.stream);
+        self.cache.deinit();
+
+        self.sync();
+
+        cuda.deinitCublasHandle(self.context.cublas);
+        cuda.deinitStream(self.context.stream);
+        self.* = undefined;
+    }
+
+    pub fn reference(self: *CudaDevice) DeviceReference {
+        return .{ .ptrs = DeviceReference.DevicePtrs{ .aux = self }, .allocator = self.allocator };
+    }
+
+    pub fn memAllocUncached(self: *CudaDevice, comptime T: type, n: usize) Error![]T {
```

Comment:
I'm adding the prefix `mem` here because it distinguishes these functions from the `allocator` functions like `alloc` and `dupe` which cannot be used interchangeably.